### PR TITLE
[Feature] bookmark

### PR DESCRIPTION
**Pull request by hanyulo**
## Features 
Bookmark 

## Targeted PR Branch
registration

## Related Repos
• [twreporter-react-components](https://github.com/twreporter/twreporter-react-components)
• [twreporter-redux](https://github.com/twreporter/twreporter-redux)
• [twreporter-registration](https://github.com/twreporter/react-redux-registration)
• [twreporter-react](https://github.com/twreporter/twreporter-react)

## TO DO 
1. Update twreporter-redux version number
2. Yarn.lock 
3. CHANGE.log
4. Publish

## Reviewer 
@nickhsine 

## Status
Wait for Sec Review

### CODE REVIEW

File: src/actions/bookmarks.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+          resolve({ data, total })
+        })
+        .catch((error) => {
+          const errorInfo = extractErrorInfo(error)
+          dispatch(errorHandler(errorInfo))
+          reject()
+        })
+    })
+  }
+}
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: `axios` returns `Promise`. 
```
return axiosInstance.get(url).then(
 ...
 return {data, total}
).catch(
  ...
)
```
BTW, add timeout restriction.

File: src/actions/bookmarks.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  } else if (error.request) {
+    // The request was made but no response was received
+    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
+    // http.ClientRequest in node.js
+    errorObj.request = _.get(error, 'request', 'default request')
+  } else {
+    // Something happened in setting up the request that triggered an Error
+    errorObj.preRequest = _.get(error, 'message', 'default message for preRequest')
+  }
+  return errorObj
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: Why do we have `error.request` ?

File: src/reducers/bookmarks.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      return {
+        ...state,
+        error: {
+          ...action.payload,
+        },
+      }
+    }
+    default:
+      return state
+  }
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/reducers/bookmarks.js]: Yes, there is only single case exist.

File: src/actions/bookmarks.js (hunk 1 above)
Comment (nickhsine) [reply src/actions/bookmarks.js]: Remember to add timeout restriction.

File: src/actions/bookmarks.js [hunk 4]
```diff
@@ -54,91 +35,79 @@ export function createBookmark({ slug, host, is_external, title, desc, thumbnail
...
-          const errorInfo = extractErrorInfo(error)
-          dispatch(errorHandler(errorInfo))
-          reject()
-        })
-    })
+    const path = `${apiEndpoints.users}/${userId}/${apiEndpoints.bookmarks}?offset=${offset}&limit=${limit}&sort=${sort}`
+    const url = formAPIURL(path)
+    return axiosInstance.get(url)
+      .then((res) => {
+        const data = _.get(res, 'data.records', '')
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: should default value be empty array?

File: src/actions/bookmarks.js [hunk 5]
```diff
@@ -54,91 +35,79 @@ export function createBookmark({ slug, host, is_external, title, desc, thumbnail
...
-        .catch(() => {
-          reject()
-        })
-    })
+    const path = `${apiEndpoints.users}/${userId}/${apiEndpoints.bookmarks}/${bookmarkSlug}?host=${host}`
+    const url = formAPIURL(path)
+    return axiosInstance.get(url)
+      .then((res) => {
+        const { id } = _.get(res, 'data.record', '')
+        return id
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: `return _.get(res, 'data.record', '')`

File: src/reducers/bookmarks.js [hunk 6]
```diff
@@ -10,9 +10,7 @@ export default function (state = initialState, action) {
     case types.RECORD_BOOKMARK_ERROR: {
       return {
         ...state,
```
Comment (nickhsine) [inline src/reducers/bookmarks.js]: why needs `...state` ?

File: src/actions/bookmarks.js [hunk 7]
```diff
@@ -54,91 +35,79 @@ export function createBookmark({ slug, host, is_external, title, desc, thumbnail
...
-          reject()
-        })
-    })
+    const path = `${apiEndpoints.users}/${userId}/${apiEndpoints.bookmarks}?offset=${offset}&limit=${limit}&sort=${sort}`
+    const url = formAPIURL(path)
+    return axiosInstance.get(url)
+      .then((res) => {
+        const data = _.get(res, 'data.records', '')
+        const total = _.get(res, 'data.meta.total', '')
+        return { data, total }
```
Comment (nickhsine) [inline src/actions/bookmarks.js]: Why do not you store `data` and  `total` in the redux state?
Do you want to fetch bookmark list every time when the user is on the bookmark list page?

File: src/constants/redux-state-field-names.js [hunk 8]
```diff
@@ -61,6 +61,7 @@ const lists = 'lists'
 const topicList = 'topic_list'
 const selectedPost = 'selected_post'
 const selectedTopic = 'selected_topic'
+const bookmarks = 'bookmarks'
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/constants/redux-state-field-names.js]: ddd

File: src/actions/bookmarks.js (hunk 4 above)
Comment (hanyulo) [reply src/actions/bookmarks.js]: 👍

File: src/actions/bookmarks.js (hunk 7 above)
Comment (hanyulo) [reply src/actions/bookmarks.js]: Fixed it. Now, users will fetch data from redux state first.

File: src/actions/bookmarks.js (hunk 5 above)
Comment (hanyulo) [reply src/actions/bookmarks.js]: 👍

File: src/constants/redux-state-field-names.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/constants/redux-state-field-names.js]: 👍

File: src/reducers/bookmarks.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/reducers/bookmarks.js]: Latest Update: For now, there are four cases in bookmark reducer.
1. Record Error
2. Get Multiple Bookmarks: 
3. Get Single Bookmark: 
4. Create Single Bookmark

**Comment by nickhsine** [comment]
Do you write the unit test for bookmark actions and reducers?

File: src/actions/bookmarks.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
+import DateStringToTimeStamp from '../utils/date-string-to-timestamp'
+import apiConfig from '../conf/api-config.json'
+import apiEndpoints from '../constants/api-endpoints'
+import axios from 'axios'
+import formAPIURL from '../utils/form-api-url'
+import { getNecessaryInfo } from '../utils/local-storage-manager'
+import types from '../constants/action-types'
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/actions/bookmarks.js]: 👍

File: src/actions/bookmarks.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    type: types.GET_MULTIPLE_BOOKMARKS,
+    payload: {
+      offset,
+      data,
+      total,
+    },
+  }
+}
+
+const createsingleBookmarkSuc = (bookmark) => {
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/actions/bookmarks.js]: 👍

File: src/actions/bookmarks.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+      .then((res) => {
+        const data = _.get(res, 'data.records', [])
+        const total = _.get(res, 'data.meta.total', 0)
+        if (data && total) {
+          dispatch(getMultipleBookmarksSuc(offset, data, total))
+        }
+        return { data, total }
+      })
+      .catch((error) => {
+        dispatch(errorHandler(error))
```
[the remark this answers is not in the document]
Comment (hanyulo) [reply src/actions/bookmarks.js]: 👍

**Comment by hanyulo** [comment]
I will write test files after I finish Sherry's project.

**Pull request merged**